File: Xit/XTRepository+Commands.m
#import "XTRepository+Commands.h"
#import "XTConstants.h"
#import "Xit-Swift.h"
#import <ObjectiveGit/ObjectiveGit.h>

@interface XTRepository()

@property (readwrite) XTRepositoryWatcher *repoWatcher;
@property (readwrite) XTWorkspaceWatcher *workspaceWatcher;
@property (readwrite) XTConfig *config;

@end


@implementation XTRepository (Commands)

- (BOOL)initializeRepository
{
  NSError *error = nil;
  GTRepository *newRepo = [GTRepository
      initializeEmptyRepositoryAtFileURL:_repoURL
                                 options:nil
                                   error:&error];
  
  if ((newRepo == nil) || (error != nil))
    return NO;
  _gtRepo = newRepo;
  self.repoWatcher = [[XTRepositoryWatcher alloc] initWithRepository:self];
  self.workspaceWatcher =
      [[XTWorkspaceWatcher alloc] initWithRepository:self];
  self.config = [[XTConfig alloc] initWithRepository:self];
  return YES;
}

- (BOOL)saveStash:(NSString*)name includeUntracked:(BOOL)untracked
{
  NSError *error = nil;
  BOOL result = NO;
  NSMutableArray<NSString*> *args =
      [NSMutableArray arrayWithObjects:@"stash", @"save", nil];

  if (untracked)
    [args addObject:@"--include-untracked"];
  if (name.length > 0)
    [args addObject:name];
  [self executeGitWithArgs:args writes:YES error:&error];

  if (error == nil) {
    result = YES;
  }

  return result;
}

- (BOOL)push:(NSString *)remote
{
  NSError *error = nil;
  BOOL result = NO;

  [self executeGitWithArgs:@[ @"push", @"--all", @"--force", remote ]
                    writes:NO
                     error:&error];

  if (error == nil) {
    result = YES;
  }

  return result;
}

- (BOOL)checkout:(NSString *)branch error:(NSError **)resultError
{
  return [self executeWritingBlock:^BOOL{
    _cachedBranch = nil;
    _cachedHeadRef = nil;
    _cachedHeadSHA = nil;

    GTCheckoutOptions *options = [GTCheckoutOptions checkoutOptionsWithStrategy:GTCheckoutStrategySafe];
    NSString *branchRef = [[GTBranch localNamePrefix]
        stringByAppendingPathComponent:branch];
    GTReference *ref = [_gtRepo lookUpReferenceWithName:branchRef error:resultError];

    if (ref != nil)
      return [_gtRepo checkoutReference:ref
                               options:options
                                 error:resultError];

    // TODO: Make a separate checkoutSHA method to avoid ambiguity
    if (branch.length == 40) {
      if (resultError != NULL)
        *resultError = nil;

      GTCommit *commit = [_gtRepo
          lookUpObjectBySHA:branch
          objectType:GTObjectTypeCommit
          error:resultError];

      if (commit != nil)
        return [_gtRepo checkoutCommit:commit
                               options:options
                                 error:resultError];
    }
    return NO;
  }];
}

- (BOOL)createTag:(NSString*)name
        targetSHA:(NSString*)sha
          message:(NSString*)message
{
  return [self executeWritingBlock:^BOOL{
    NSError *error = nil;
    GTCommit *targetCommit = [_gtRepo lookUpObjectBySHA:sha
                                             objectType:GTObjectTypeCommit
                                                  error:&error];
    GTSignature *signature = [_gtRepo userSignatureForNow];

    if ((targetCommit == nil) || (signature == nil))
      return NO;

    [_gtRepo createTagNamed:name
                     target:targetCommit
                     tagger:[_gtRepo userSignatureForNow]
                    message:message
                      error:&error];
    return error == nil;
  }];
}

- (BOOL)createLightweightTag:(NSString*)name targetSHA:(NSString*)sha
{
  return [self executeWritingBlock:^BOOL{
    NSError *error = nil;
    GTCommit *targetCommit = [_gtRepo lookUpObjectBySHA:sha
                                             objectType:GTObjectTypeCommit
                                                  error:&error];
    
    if (targetCommit == nil)
      return NO;
    [_gtRepo createLightweightTagNamed:name
                                target:targetCommit
                                 error:&error];
    return error != nil;
  }];
}

- (BOOL)deleteTag:(NSString *)name error:(NSError *__autoreleasing *)error
{
  return [self executeWritingBlock:^BOOL{
    int result = git_tag_delete([_gtRepo git_repository], name.UTF8String);

    if (result == 0)
      return YES;
    else {
      if (error != NULL)
        *error = [NSError git_errorFor:result];
      return NO;
    }
  }];
}

- (BOOL)addRemote:(NSString *)name withUrl:(NSString *)url
{
  NSError *error = nil;
  BOOL result = NO;

  [self executeGitWithArgs:@[ @"remote", @"add", name, url ]
                    writes:YES
                     error:&error];

  if (error == nil) {
    result = YES;
  }

  return result;
}

- (BOOL)deleteRemote:(NSString *)name error:(NSError *__autoreleasing *)error
{
  return [self executeGitWithArgs:@[ @"remote", @"rm", name ]
                           writes:YES
                            error:error] != nil;
}

- (NSString *)urlStringForRemote:(NSString *)remoteName
{
  NSString *remoteURL =
      [[_gtRepo configurationWithError:nil] stringForKey:remoteName];
  
  return remoteURL;
}

- (BOOL)stagePatch:(NSString *)patch
{
  NSError *error = nil;

  [self executeGitWithArgs:@[ @"apply", @"--cached" ]
                 withStdIn:patch
                    writes:YES
                     error:&error];
  return error == nil;
}

- (BOOL)unstagePatch:(NSString *)patch
{
  NSError *error = nil;

  [self executeGitWithArgs:@[ @"apply", @"--cached", @"--reverse" ]
                 withStdIn:patch
                    writes:YES
                     error:&error];
  return error == nil;
}

- (BOOL)discardPatch:(NSString *)patch
{
  NSError *error = nil;

  [self executeGitWithArgs:@[ @"apply", @"--reverse" ]
                 withStdIn:patch
                    writes:YES
                     error:&error];
  return error == nil;
}

- (BOOL)unstageAllFiles
{
  // "Unstaging" just means making things match what's in the head commit,
  // so if we're unstaging everything we can just copy the tree over.
  NSError *error = nil;
  GTReference *headRef = [_gtRepo headReferenceWithError:&error];
  GTCommit *headCommit = (GTCommit*)headRef.resolvedTarget;
  
  if ((headCommit == nil) || ![headCommit isKindOfClass:[GTCommit class]])
    return NO;
  
  GTIndex *index = [_gtRepo indexWithError:&error];
  
  if (index == nil)
    return NO;
  
  if (![index addContentsOfTree:headCommit.tree error:&error]) {
    NSLog(@"couldn't unstage all: %@", error);
    return NO;
  }
  if (![index write:&error]) {
    NSLog(@"couldn't write index: %@", error);
    return NO;
  }
  return YES;
}

- (BOOL)renameRemote:(NSString *)branch to:(NSString *)newName
{
  NSError *error = nil;

  [self executeGitWithArgs:@[ @"remote", @"rename", branch, newName ]
                    writes:YES
                     error:&error];
  return error == nil;
}

- (GTCheckoutOptions*)stashCheckoutOptions
{
  return [GTCheckoutOptions checkoutOptionsWithStrategy:GTCheckoutStrategySafe];
}

- (BOOL)popStashIndex:(NSUInteger)index error:(NSError**)error
{
  return [self executeWritingBlock:^BOOL{
    return [self.gtRepo popStashAtIndex:index
                                  flags:GTRepositoryStashApplyFlagReinstateIndex
                        checkoutOptions:[self stashCheckoutOptions]
                                  error:error
                          progressBlock:nil];
  }];
}

- (BOOL)applyStashIndex:(NSUInteger)index error:(NSError**)error
{
  return [self executeWritingBlock:^BOOL{
    return [self.gtRepo applyStashAtIndex:index
                                    flags:GTRepositoryStashApplyFlagReinstateIndex
                          checkoutOptions:[self stashCheckoutOptions]
                                    error:error
                            progressBlock:nil];
  }];
}

- (BOOL)dropStashIndex:(NSUInteger)index error:(NSError**)error
{
  return [self executeWritingBlock:^BOOL{
    return [self.gtRepo dropStashAtIndex:index error:(NSError**)error];
  }];
}

- (BOOL)addSubmoduleAtPath:(NSString *)path
                 urlOrPath:(NSString *)urlOrPath
                     error:(NSError **)error
{
  return [self executeGitWithArgs:@[ @"submodule", @"add", @"-f",
                                     urlOrPath, path ]
                           writes:YES
                            error:error] != nil;
/* The clone step must be implemented for this to be good.
  return [self executeWritingBlock:^BOOL{
    git_submodule *gitSub = NULL;
    int result = git_submodule_add_setup(
        &gitSub, [gtRepo git_repository],
        [urlOrPath UTF8String], [path UTF8String], false);

    if ((result != 0) && (error != NULL)) {
      *error = [NSError git_errorFor:result];
      return NO;
    }
    // clone the sub-repo
    git_submodule_add_finalize(gitSub);
    return YES;
  }];
*/
}

@end

File: Xit/XTRepository+Commands.swift
import Foundation

extension XTRepository
{
  func createTag(name: String, targetOID: OID, message: String?) throws
  {
    try performWriting {
      guard let commit = XTCommit(oid: targetOID,
                                  repository: gtRepo.git_repository())
      else { throw Error.notFound }
      
      let oid = UnsafeMutablePointer<git_oid>.allocate(capacity: 1)
      let signature = UnsafeMutablePointer<UnsafeMutablePointer<git_signature>?>
            .allocate(capacity: 1)
      let sigResult = git_signature_default(signature, gtRepo.git_repository())
      
      try Error.throwIfError(sigResult)
      guard let finalSig = signature.pointee
      else { throw Error.unexpected }
      
      let result = git_tag_create(oid, gtRepo.git_repository(), name,
                                  commit.commit, finalSig, message, 0)
      
      try Error.throwIfError(result)
    }
  }
  
  func createLightweightTag(name: String, targetOID: OID) throws
  {
    try performWriting {
      guard let commit = XTCommit(oid: targetOID,
                                  repository: gtRepo.git_repository())
      else { throw Error.notFound }
      
      let oid = UnsafeMutablePointer<git_oid>.allocate(capacity: 1)
      let result = git_tag_create_lightweight(oid, gtRepo.git_repository(), name,
                                              commit.commit, 0)
      
      try Error.throwIfError(result)
    }
  }
  
  func deleteTag(name: String) throws
  {
    try performWriting {
      let result = git_tag_delete(gtRepo.git_repository(), name)
      
      guard result == 0
      else {
        throw NSError.git_error(for: result)
      }
    }
  }
  
  func push(remote: String) throws
  {
    _ = try executeGit(args: ["push", "--all", remote], writes: true)
  }
  
  func checkout(branch: String) throws
  {
    try performWriting {
      // invalidate ref caches
      
      let branchRef = BranchPrefixes.heads.appending(pathComponent: branch)
      let ref = try gtRepo.lookUpReference(withName: branchRef)
      let options = GTCheckoutOptions(strategy: [.safe])
      
      try gtRepo.checkoutReference(ref, options: options)
    }
  }
  
  func checkout(sha: String) throws
  {
    guard let commit = try gtRepo.lookUpObject(bySHA: sha) as? GTCommit
    else { throw Error.unexpected }
    let options = GTCheckoutOptions(strategy: [.safe])
    
    try gtRepo.checkoutCommit(commit, options: options)
  }
  
  func stagePatch(_ patch: String) throws
  {
    _ = try executeGit(args: ["apply", "--cached"], stdIn: patch, writes: true)
  }
  
  func unstagePatch(_ patch: String) throws
  {
    _ = try executeGit(args: ["apply", "--cached", "--reverse"],
                       stdIn: patch,
                       writes: true)
  }
  
  func discardPatch(_ patch: String) throws
  {
    _ = try executeGit(args: ["apply", "--reverse"],
                       stdIn: patch,
                       writes: true)
  }
  
  func unstageAllFiles() throws
  {
    guard let index = GitIndex(repository: self)
    else { throw Error.unexpected }
    
    if let headOID = headReference?.resolve()?.targetOID {
      guard let headCommit = commit(forOID: headOID),
            let headTree = headCommit.tree
      else { throw Error.unexpected }
      
      try index.read(tree: headTree)
    }
    else {
      // If there is no head, then this is the first commit
      try index.clear()
    }

    try index.save()
  }
  
  func renameRemote(old: String, new: String) throws
  {
    _ = try executeGit(args: ["remote", "rename", old, new], writes: true)
  }
}

extension XTRepository: Stashing
{
  // TODO: Don't require the message parameter
  public func stash(index: UInt, message: String?) -> Stash
  {
    return XTStash(repo: self, index: index, message: message)
  }
  
  @objc(saveStash:includeUntracked:error:)
  func saveStash(name: String?, includeUntracked: Bool) throws
  {
    var args = ["stash", "save"]
    
    if includeUntracked {
      args.append("--include-untracked")
    }
    if let name = name {
      args.append(name)
    }
    _ = try executeGit(args: args, stdIn: nil, writes: true)
  }
  
  func stashCheckoutOptions() -> GTCheckoutOptions
  {
    return GTCheckoutOptions(strategy: .safe)
  }
  
  public func popStash(index: UInt) throws
  {
    _ = try performWriting {
      try gtRepo.popStash(at: index, flags: [.reinstateIndex],
                          checkoutOptions: stashCheckoutOptions(),
                          progressBlock: nil)
    }
  }
  
  public func applyStash(index: UInt) throws
  {
    _ = try performWriting {
      try gtRepo.applyStash(at: index, flags: [.reinstateIndex],
                            checkoutOptions: stashCheckoutOptions(),
                            progressBlock: nil)
    }
  }
  
  public func dropStash(index: UInt) throws
  {
    _ = try performWriting {
      try gtRepo.dropStash(at: index)
    }
  }
  
  public func commitForStash(at index: UInt) -> Commit?
  {
    guard let stashRef = try? gtRepo.lookUpReference(withName: "refs/stash"),
          let stashLog = GTReflog(reference: stashRef),
          index < stashLog.entryCount,
          let entry = stashLog.entry(at: index),
          let oid = entry.updatedOID.map({ GitOID(oid: $0.git_oid().pointee) })
    else { return nil }
    
    return XTCommit(oid: oid, repository: gtRepo.git_repository())
  }
}

extension XTRepository: RemoteManagement
{
  public func remote(named name: String) -> Remote?
  {
    return GitRemote(name: name, repository: gtRepo.git_repository())
  }
  
  public func addRemote(named name: String, url: URL) throws
  {
    _ = try executeGit(args: ["remote", "add", name, url.absoluteString],
                       writes: true)
  }
  
  public func deleteRemote(named name: String) throws
  {
    _ = try executeGit(args: ["remote", "rm", name],
                       writes: true)
  }
}

extension XTRepository: SubmoduleManagement
{
  public func addSubmodule(path: String, url: String) throws
  {
    _ = try executeGit(args: ["submodule", "add", "-f", url, path],
                       writes: true)
    /* still needs clone
    _ = try performWriting {
     git_submodule *gitSub = NULL;
     let result = git_submodule_add_setup(
        &gitSub, [gtRepo git_repository],
        [urlOrPath UTF8String], [path UTF8String], false);
     
     if ((result != 0) && (error != NULL)) {
      *error = [NSError git_errorFor:result];
      return NO;
     }
     // clone the sub-repo
     git_submodule_add_finalize(gitSub);
    }
    */
  }
  
  public func submodules() -> [Submodule]
  {
    class Payload { var submodules = [Submodule]() }
    var payload = Payload()
    let callback: git_submodule_cb = {
      (submodule, name, payload) in
      guard let submodule = submodule,
            let repo = git_submodule_owner(submodule)
      else { return 0 }
      let payload = payload!.bindMemory(to: Payload.self, capacity: 1)
      
      // Look it up again to get an owned reference
      let mySubmodule = UnsafeMutablePointer<OpaquePointer?>.allocate(capacity: 1)
      let lookup = git_submodule_lookup(mySubmodule, repo,
                                        git_submodule_name(submodule))
      guard lookup == 0,
            let finalSubmodule = mySubmodule.pointee
      else { return 0 }
      
      payload.pointee.submodules.append(GitSubmodule(submodule: finalSubmodule))
      return 0
    }
    
    git_submodule_foreach(gtRepo.git_repository(), callback, &payload)
    return payload.submodules
  }
}
